posthog/temporal/workflows/squash_person_overrides.py

```diff
@@ -0,0 +1,698 @@
+import json
+import os
+from collections.abc import Iterator
+from dataclasses import asdict, dataclass, field
+from datetime import datetime, timedelta, timezone
+from typing import Iterable, NamedTuple
+from uuid import UUID
+
+import psycopg2
+from django.conf import settings
+from temporalio import activity, workflow
+from temporalio.common import RetryPolicy
+
+from posthog.clickhouse.client.execute import sync_execute
+from posthog.temporal.workflows.base import CommandableWorkflow
+
+EPOCH = datetime(1970, 1, 1, 0, 0, tzinfo=timezone.utc)
+
+SELECT_PERSONS_TO_DELETE_QUERY = """
+SELECT
+    team_id,
+    old_person_id,
+    override_person_id,
+    max(created_at) AS latest_created_at,
+    max(version) AS latest_version,
+    min(oldest_event) AS oldest_event_at
+FROM
+    {database}.person_overrides
+WHERE
+    created_at <= %(latest_created_at)s
+GROUP BY
+    team_id, old_person_id, override_person_id
+"""
+
+SELECT_LATEST_CREATED_AT_QUERY = """
+SELECT
+    max(created_at)
+FROM {database}.person_overrides;
+"""
+
+CREATE_DICTIONARY_TABLE_QUERY = """
+CREATE OR REPLACE TABLE {database}.{dictionary_name}_table (
+    team_id INT NOT NULL,
+    old_person_id UUID NOT NULL,
+    override_person_id UUID NOT NULL
+)
+ENGINE = TinyLog();
```

Comment:
A few points about the TinyLog engine:

- Not replicated
- Not efficient for point queries ([docs](https://clickhouse.com/docs/en/engines/table-engines/log-family#common-properties))

We're also not creating it on cluster here so this won't work as the source for the dictionary on all nodes.

I see what we're trying to do here - be as efficient as possible with a temporary table. But easiest/best is to just create a standard merge tree table at the end of the day.

Create it replicated on the whole cluster and then run the insert only on one node. Probably need to add some "wait for replication" step though

Reply:
Thinking here now, there's maybe a simpler option.

`person_overrides` uses the ReplacingMergeTree engine, which was a great decision! Essentially we can just force optimize that table (`OPTIMIZE TABLE person_overrides FINAL`) and that'll get rid of duplicates for us. 

Also another thing we need to do in general here which I just remembered is we need to stop ingesting to `person_overrides`, i.e. detach `person_overrides` on cluster before starting and then re-attaching it